apex/contrib/torchsched/ops/layer_norm.py

```diff
@@ -0,0 +1,468 @@
+"""Customized CuDNN frontend layer norm.
+
+Please refer to:
+
+* https://github.com/NVIDIA/cudnn-frontend/blob/main/samples/python/20_layernorm.ipynb
+* https://gitlab-master.nvidia.com/dl/mlperf/optimized/-/merge_requests/4700
+"""
+
+from __future__ import annotations
+
+import math
+from typing import Sequence
+
+import cudnn
+import torch
+
+__all__ = ["get_cudnn_manager"]
+
+
+class CuDNNManager:
+    """CuDNN fronted context manager.
+
+    Notice: CuDNN handle must be created after distributed process group initialization.
+    """
+
+    def __init__(self) -> None:
+        self._handle = cudnn.create_handle()
+        self._cudnn_stream = torch.cuda.Stream()
+        self.reset_stream()
+
+    def __del__(self) -> None:
+        if cudnn is not None and hasattr(cudnn, "destroy_handle"):
+            cudnn.destroy_handle(self._handle)
+
+    def __enter__(self) -> CuDNNManager:
+        self._torch_stream = torch.cuda.current_stream()
+        self._cudnn_stream.wait_stream(self._torch_stream)
+        torch.cuda.set_stream(self._cudnn_stream)
+        return self
+
+    def __exit__(
+        self,
+        exc_type: type | None,
+        exc_val: Exception | None,
+        exc_tb: object | None,
+    ) -> None:
+        self._torch_stream.wait_stream(self._cudnn_stream)
+        torch.cuda.set_stream(self._torch_stream)
+        del self._torch_stream
+
+    def set_stream(self, stream: torch.cuda.Stream) -> None:
+        cudnn.set_stream(stream=stream.cuda_stream, handle=self._handle)
+
+    def reset_stream(self) -> None:
+        cudnn.set_stream(stream=self._cudnn_stream.cuda_stream, handle=self._handle)
+
+    @property
+    def handle(self) -> int:
+        return self._handle
+
+    @property
+    def stream(self) -> torch.cuda.Stream:
+        return self._cudnn_stream
+
+
+_global_cudnn_manager: CuDNNManager | None = None
+
+
+def get_cudnn_manager() -> CuDNNManager:
+    """Get the CuDNN front-end context manager.
+
+    Returns:
+        CuDNNManager: Global CuDNN manager.
+    """
+    global _global_cudnn_manager
+    if _global_cudnn_manager is None:
+        _global_cudnn_manager = CuDNNManager()
+    return _global_cudnn_manager
+
+
+class LayerNormGraphFactory:
```

Comment:
Have you evaluated cuDNN layernorm performance vs. native PyTorch layer norm? In my testing I've found that in many cases, especially for smaller shapes native PyTorch is faster.

Reply:
That's interesting, native layer norm perfoms quite worse for our cases, especially on B200.